### PR TITLE
Infra for tvm op runtime dispatch

### PR DESCRIPTION
**Pull request by hzfan**
## Description ##
This PR implements an infra to let users dispatch the execution of a tvm operator to different schedules according to the runtime input shapes. This helps with acceleration. 

A gemm example can be found in
- Kernel definition: contrib/tvmop/core/multiarray.py
- Operator registry and dispatch: src/operator/contrib/tvmop/dot.cc
- Benchmark: benchmark/python/tvmop/benchmark_tvmop.py

The following are some experimental results for matrix multiplication between two n * n matrix. Note that benchmark results cannot be reproduced until [this](https://github.com/dmlc/tvm/pull/3842) gets merged.

| n  | After Dispatch (ms) | Before Dispatch (ms) |
| --- | --- | --- |
| 1024  | 177 | 482 |
| 1056  | 190 | 366 |
| 1088 | 200 | 424 |

The example schedule is roughly equivalent to the [Blocking](https://docs.tvm.ai/tutorials/optimize/opt_gemm.html) optimization. More opt (like vectorization, loop permutation, array packing, write cache for blocks, parallel) can be used for further acceleration.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- Add dispatch infra
- Add an example dot operator

## Comments ##
- Thank @yzhliu and @reminisce for guidance and review.

### CODE REVIEW

**Comment by ZhennanQin** [comment]
Just for curious. Based on my knowledge, tvm op kernel is pre-compiled and then linked together with MXNet. How can it be configured according to the runtime input shapes?

**Comment by hzfan** [comment]
> Just for curious. Based on my knowledge, tvm op kernel is pre-compiled and then linked together with MXNet. How can it be configured according to the runtime input shapes?

Yes, kernels are pre-compiled. At compile time, several different schedules (kernels) for a single op are defined and compiled. Then at runtime, with the runtime input shape, the most suitable kernel is chosen.

It's true that the kernel is pre-compiled, but we have multiple available kernels for one single op, so we can choose the most efficient one based on the runtime input shape.

**Comment by ZhennanQin** [comment]
@hzfan Thanks for explanation.  My next question is, how do I know which schedule is the best one for a certain input shape? Static rule defined or runtime tuned?

**Comment by hzfan** [comment]
> @hzfan Thanks for explanation. My next question is, how do I know which schedule is the best one for a certain input shape? Static rule defined or runtime tuned?

That's a good question. Actually we have considered both options, and for now we use simple static rules. To be more specific, for now I require the size of a for-loop to be multiples of its splitting factor (if the for-loop is splitted). This helps eliminate a if-condition, and thus makes it faster.

Runtime tuning has also been considered, but has not been implemented in this version. The idea is to try all the available schedules for every runtime shape, measure their performance, and cache the best choice. This is quite similar to autotvm.

**Comment by yzhliu** [comment]
Also cc @icemelon9 @kevinthesun

File: benchmark/python/tvmop/benchmark_tvmop.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    start = time.time()
+    for _ in range(repeat):
+        func_name(*args, **kwargs)
+    mx.nd.waitall()
+    end = time.time()
+    diff = end - start
+    return diff / repeat
+
+
+def test_tvm_dot():
```
[the remark this answers is not in the document]
Comment (hzfan) [reply benchmark/python/tvmop/benchmark_tvmop.py]: Yes, for reproducing the benchmark result. Other code in `benchmark/` only serves this purpose, too.

File: python/mxnet/libinfo.py [hunk 2]
```diff
@@ -110,5 +110,36 @@ def find_include_path():
...
+
+    curr_path = os.path.dirname(os.path.abspath(os.path.expanduser(__file__)))
+    makefile_path = os.path.join(curr_path, '../../')
+    cmake_build_path = os.path.join(curr_path, '../../build/')
+    candidates_path = [makefile_path, cmake_build_path]
+    candidates_path = [p + prefix + '.conf' for p in candidates_path]
+    conf_path = [p for p in candidates_path if os.path.exists(p) and os.path.isfile(p)]
+    if len(conf_path) == 0:
+        raise RuntimeError('Cannot find the TVM op config.\n' +
+                           'List of candidates:\n' + str('\n'.join(candidates_path)))
```
[the remark this answers is not in the document]
Comment (hzfan) [reply python/mxnet/libinfo.py]: For now we can't, but it's doable.

In which case will the config file be missing? It is generated in compile time (even if no tunable parameters are needed, a nearly empty config will be generated too).

File: contrib/tvmop/compile.py [hunk 3]
```diff
@@ -37,23 +40,57 @@ def get_target(device):
...
+        if tvm.module.enabled(get_target(op.target)):
+            func_list = func_list_llvm if op.target == "cpu" else func_list_cuda
+            for each_kwargs in op.arg_combination:
+                if (op.attrs_valid(**each_kwargs)):
+                    name = op.name \
+                        + ''.join(["{}_{}".format(key, each_kwargs[key]) for key in op.attrs])
+                    if op.dispatch is True:
+                        config_space = autotvm.ConfigSpace()
+                        with autotvm.task.ApplyConfig(config_space):
+                            sch, args = op.func(fallback=False, **each_kwargs)
```
Comment (reminisce) [inline contrib/tvmop/compile.py]: This requires `fallback` as a mandatory parameter in `op.func`, which is not ideal in terms of usability in my opinion. We should support compiling whatever users define and treat the `fallback` knob as an advanced feature for performance tuning.

A way to achieve such purpose is inspect the signature of `op.func` for keyword `fallback`. If the keyword does not exist, we just compile the op using the default schedule, e.g.
```python
if 'fallback' in str(inspect.signature(op.func)):
    sch, args = op.func(fallback=False, **each_kwargs)
else:
    sch, args = op.func(**each_kwargs)
```
@yzhliu What do you think?

File: contrib/tvmop/core/multiarray.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+                    lambda x, y: tvm.sum(A[x, k] * B[k, y], axis=k),
+                    name='C')
+    return C
+
+
+@defop(name="dot", target="cpu", dispatch=True, dtype=AllTypes)
+def dot(dtype, fallback):
+    cfg = autotvm.get_config()
+    cfg.define_knob("bn", [64] if fallback else [64, 32])
+    cfg.define_knob("factor", [4] if fallback else [4])
```
[the remark this answers is not in the document]
Comment (hzfan) [reply contrib/tvmop/core/multiarray.py]: Yes. The difference is that when `fallback` is false, the shape comes with a hint, indicating that it is multiples of 4.

This `factor` means in any case I want to split the loop by a factor of 4. When fallback, there is no guarantee the loop size is a multiples of 4., while when not fallback, there is.

File: contrib/tvmop/space.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# coding: utf-8
+"""ConfigSpace API."""
+from collections import OrderedDict
+import numpy as _np
+
+class OtherOptionSpace(object):
```
[the remark this answers is not in the document]
Comment (hzfan) [reply contrib/tvmop/space.py]: Actually the `OtherOptionSpace` comes from `tvm/python/tvm/autotvm/task/space.py`. Besides `OtherOptionSpace`, there is `SplitSpace`, `ReorderSpace` and `AnnotateSpace`. Maybe in the future the three other spaces will be needed, so I keep its name consistent with tvm.

File: contrib/tvmop/core/multiarray.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (reminisce) [reply contrib/tvmop/core/multiarray.py]: I understand what you are trying to get here. My point is that this line is equivalent to the following, correct?
```python
cfg.define_knob("factor", [4])
```

File: contrib/tvmop/space.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (reminisce) [reply contrib/tvmop/space.py]: Fair enough.

File: src/operator/tvmop/op_module.h [hunk 6]
```diff
@@ -56,6 +57,79 @@ class TVMOpModule {
...
+  inline const OtherOptionSpace& get_space(const std::string& name) const {
+    return space_map.at(name);
+  }
+
+  inline int get_weight(const std::string& name) const {
+    return weight_map.at(name);
+  }
+
+ private:
+  std::map<std::string, OtherOptionEntity> _entity_map;
```
[the remark this answers is not in the document]
Comment (hzfan) [reply src/operator/tvmop/op_module.h]: Fixed.

File: contrib/tvmop/core/multiarray.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hzfan) [reply contrib/tvmop/core/multiarray.py]: Yes.

File: contrib/tvmop/compile.py (hunk 3 above)
Comment (hzfan) [reply contrib/tvmop/compile.py]: Done. I set `self.dispatchable = 'fallback' in inspect.signature(self.func).parameters` in `opdef.py`

**Comment by eric-haibin-lin** [comment]
do we have a developer guide using tvm op?

**Comment by hzfan** [comment]
> do we have a developer guide using tvm op?

Seems @yzhliu is working on it.

**Pull request merged**